### PR TITLE
Type changing lenses compatible with nominal roles

### LINKED ISSUE #55: Type-changing lenses with nominal roles: a "field" without type inference
Follow up on #40, #45

It seems that for the `field` lens, the type parameter tagging is used only for type inference, and the above issues mention it breaking in the presence of nominal roles. What do you think of a `HasField0` class that is a variant of `HasField`, without type inference tricks, so that it can still be used with more types? In fact, the `HasField` instance can then be expressed in terms of `HasField0`.

---

I have a quick proof of concept disabling type inference here

https://github.com/Lysxia/generic-lens/commit/3b20812efa93274421deaffb64cf2168635cb372

So we can indeed use type-changing lenses with data families:

```
> import Data.Generics.Product
> import Data.Generics.Internal.VL.Lens 
> data family T a :: * -> *
> data instance T Bool b = C { f :: b } deriving (Show,Generic)
> field @"f" @_ @(T Bool Int) .~ 3 $ C ()     -- T Bool () -> T Bool Int
C {f = 3}
```

---

On a related note, have you considered using the following class instead of the "tagging trick" for unification of `s` and `t`:

```
class Unify (a :: k) (b :: k)
instance {-# OVERLAPPING #-} (gb ~ g b, Unify f g) => Unify (f a) gb
instance (a ~ b) => Unify a b
```

This unifies just the type constructor,  which is not as good as the tagging trick in general, but is often good enough for `Rep` to reduce and let the other constraints do their job. It might also work well with the type family example of #45 if for some reason you can otherwise infer the parameter `a` but not the type constructor.

Together with my proposal above that could be just another variant of `field` to export.

### PR DESCRIPTION
**Pull request by Lysxia**
Closes #55.

### CODE REVIEW

File: src/Data/Generics/Product/Fields.hs [hunk 1]
```diff
@@ -159,6 +168,18 @@ instance  -- see Note [Changing type parameters]
 instance {-# OVERLAPPING #-} HasField f (Void1 a) (Void1 b) a b where
   field = undefined
 
+instance
+  ( Generic s
+  , ErrorUnless field s (CollectField field (Rep s))
+  , Generic t
+  , GHasKey  field (Rep s) (Rep t) a b
+  , UnifyHead s t
+  ) => HasField_ field s t a b where
```
Comment (Lysxia) [inline src/Data/Generics/Product/Fields.hs]: Should I keep the `GHasKey' field (Rep s) a` from the `HasField` instance? What do those `GHasKey'` constraints do?

**Comment by Lysxia** [comment]
That sounds like a good plan to me!

**Comment by Lysxia** [comment]
Actually, were you saying this PR is a breaking change, because it's only defining a new class (see my latest comments in #55).

**Comment by kcsongor** [comment]
Right, sorry - I was thinking that it might make sense to lose those fundeps, as they seem to be a small price to pay for the simplicity of `UnifyHead`. I think that in practice type inference shouldn't suffer because that instance is general enough to be picked unless there's an overlapping one.

**Comment by Lysxia** [comment]
That makes sense!

**Comment by kcsongor** [comment]
Sorry I took quite a while to look at this again! Unfortunately it does seem like the extra functional dependencies are required when composing polymorphic lenses, such as in this scenario: https://github.com/kcsongor/generic-lens/issues/62
Incidentally, the above broke because I deleted the `GHasKey' field (Rep s) a` constraint from the lens (thinking it was redundant), but now we know why it was needed!
I tried the above with `field_`, but the intermediate types in the composition can't be resolved due to the missing fundeps.

Quite unfortunate, because this means we need two different ways for doing very similar things (at least for now). Do you have any ideas on how to best incorporate this into the API? In particular, I think the `UnifyHead` version should be used in most of the cases, and the indexing one when you need a bit better type inference.

**Comment by Lysxia** [comment]
Changing one `UnifyHead` instance from

```haskell
instance UnifyHead f g => UnifyHead (f a) (g b)
```

to

```haskell
instance (UnifyHead f g, gb ~ g b) => UnifyHead (f a) gb
```

and adding both constraints to `HasField_`, etc.:

```haskell
( UnifyHead s t
, UnifyHead t s
, ... )
```

seems to pass the test #62.

I also added `HasField0`, and both `HasField` and `HasField_` are now implemented using `HasField0`.

File: src/Data/Generics/Product/Positions.hs [hunk 2]
```diff
@@ -147,6 +165,37 @@ coerced = coerce
...
 
+instance
+  ( Generic s
+  , Generic t
+  , ErrorUnless i s (0 <? i && i <=? Size (Rep s))
+  , GLens (HasTotalPositionPSym i) (CRep s) (CRep t) a b
+  , HasTotalPositionP i (CRep s) ~~ 'Just a
+  , HasTotalPositionP i (CRep t) ~~ 'Just b
+  , HasTotalPositionP i (CRep (Indexed s)) ~~ 'Just a'
+  , HasTotalPositionP i (CRep (Indexed t)) ~~ 'Just b'
```
Comment (kcsongor) [inline src/Data/Generics/Product/Positions.hs]: Should these `Indexed` ones be removed from here? I think they're redundant (same for `HasPosition_`), given that `Infer` is no longer using them.

**Comment by kcsongor** [comment]
This looks very good, thanks for adding the documentation too! I'll find some time tomorrow to merge it in and make a release with the new changes. Thanks!

File: src/Data/Generics/Product/Positions.hs (hunk 2 above)
Comment (Lysxia) [reply src/Data/Generics/Product/Positions.hs]: Are you proposing to remove all `HasTotal*** ... (CRep (Indexed _)) ...`?

**Pull request merged**